src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs

```diff
@@ -916,6 +898,292 @@ I1<I2> M(A a)
+        }
+
+        [Fact]
+        public void RefAssemblyClient_EmitVariance_OutSuccess()
+        {
+            VerifyRefAssemblyClient(@"
+public interface I<out T>
+{
+}",
+@"
+class Base { }
+class Derived : Base
+{
+    I<Base> M(I<Derived> x)
+    {
+        return x;
+    }
+}",
+comp => comp.VerifyDiagnostics());
+        }
+
+        [Fact]
+        public void RefAssemblyClient_EmitVariance_InSuccess()
+        {
+            VerifyRefAssemblyClient(@"
+public interface I<in T>
+{
+}",
+@"
+class Base { }
+class Derived : Base
+{
+    I<Derived> M(I<Base> x)
+    {
+        return x;
+    }
+}",
+comp => comp.VerifyDiagnostics());
+        }
+
+        [Fact]
+        public void RefAssemblyClient_EmitVariance_InError()
+        {
+            VerifyRefAssemblyClient(@"
+public interface I<in T>
+{
+}",
+@"
+class Base { }
+class Derived : Base
+{
+    I<Base> M(I<Derived> x)
+    {
+        return x;
+    }
+}",
+comp => comp.VerifyDiagnostics(
+                // (7,16): error CS0266: Cannot implicitly convert type 'I<Derived>' to 'I<Base>'. An explicit conversion exists (are you missing a cast?)
+                //         return x;
+                Diagnostic(ErrorCode.ERR_NoImplicitConvCast, "x").WithArguments("I<Derived>", "I<Base>").WithLocation(7, 16)
+                ));
+        }
+
+        [Fact]
+        public void RefAssemblyClient_EmitOptionalArguments()
+        {
+            VerifyRefAssemblyClient(@"
+public class A
+{
+    public void M(int x = 42) { }
+}",
+@"
+class C
+{
+    void M2(A a)
+    {
+        a.M();
+    }
+}",
+comp =>
+{
+    comp.VerifyDiagnostics();
+    var verifier = CompileAndVerify(comp);
+    verifier.VerifyIL("C.M2", @"
+{
+  // Code size       11 (0xb)
+  .maxstack  2
+  IL_0000:  nop
+  IL_0001:  ldarg.1
+  IL_0002:  ldc.i4.s   42
+  IL_0004:  callvirt   ""void A.M(int)""
+  IL_0009:  nop
+  IL_000a:  ret
+}");
+});
+        }
+
+        [Fact]
+        public void RefAssemblyClient_EmitArgumentNames()
+        {
+            VerifyRefAssemblyClient(@"
+public class Base
+{
+    public virtual void M(int x) { }
+}
+public class Derived : Base
+{
+    public override void M(int different) { }
+}",
+@"
+class C
+{
+    void M2(Derived d)
+    {
+        d.M(different: 1);
+    }
+}",
+comp => comp.VerifyDiagnostics());
+        }
+
+        [Fact]
+        public void RefAssemblyClient_EmitEnum()
+        {
+            VerifyRefAssemblyClient(@"
+public enum E
+{
+    Default,
+    Other
+}",
+@"
+class C
+{
+    void M2(E e)
+    {
+        System.Console.Write(E.Other);
+    }
+}",
+comp => comp.VerifyDiagnostics());
+        }
+
+        [Fact]
+        public void RefAssemblyClient_EmitConst()
+        {
+            VerifyRefAssemblyClient(@"
+public class A
+{
+    public const int number = 42;
+}",
+@"
+class C
+{
+    void M2()
+    {
+        System.Console.Write(A.number);
+    }
+}",
+comp =>
+{
+    comp.VerifyDiagnostics();
+    var verifier = CompileAndVerify(comp);
```

Comment:
No, unfortunately, we cannot execute scenarios with ref assemblies because ref assemblies won't be loaded by the runtime.
I could test the scenario: (1) compile client against ref assembly, (2) execute against the real thing, but I think the IL check achieves the confidence we want (ie the ref assembly still has the constant, so the client ends up inlining that value).